### PR TITLE
Implementation of `@typeChangedFrom` decorator

### PR DESCRIPTION
**Pull request by tjprescott**
Fixes: #1372

Adds new decorators:
- `@typeChangedFrom`: works on model properties
- `@returnTypeChangedFrom`: works on operations

### CODE REVIEW

**Comment by tjprescott** [comment]
Here's a simply playground for testing this:
https://cadlplayground.z22.web.core.windows.net/prs/1408/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOw0K0xt2ZXJzaW9uaW5nxCENCnVzaW5nIENhZGwuVskbO80YSHR0cMUsQMc9ZWQoxyxzKQ0KQHNlcnZpY2Uoe25hbWU6ICJUxH59KQ0KxBBzcGFjZSDEE8VDZW51bSDIPSB7DQogIHYxLMUHMg0KfcQjbW9kZWwgRm9vxh9AdHlwZUNoYW5nZWRGcm9tyXsudjIsIHN0cmluZynEKWFkZGVkT246IHrkAKJEYXRlVGltZTvHVkBnZXQNCkByb3V0ZSgiL3TkAKQpDQpvcCDECygpOsRyOw%3D%3D

**Comment by tjprescott** [comment]
This draft implementation applies only to model properties. This allows it to work in the scenario where an operation param type changes since operation signatures are just secret models.

The other possible use case for this would be where the operation return type changes. In this case, one could imaging decorating the return type model and having validation ensure that this decorator can only be used on a model if it is a return type.

**Comment by tjprescott** [comment]
At today's design discussion, we decided to keep `typeChangedFrom` and add `returnTypeChangedFrom` which follows the same pattern but applies to operations.

We deferred discussion on potentially changing the signature of these versioning decorators that take multiple parameters.

We also need to consider how these decorators will (or will not) be used in Azure and codegen.

**Review by markcowl — APPROVED** [review]
Make sure there is a companion PR updating the azure/cadl-azure repo that is passing before you merge.

Need to file an issue to disallow use of the new decorators in the `@azure-tools/cadl-providerhub' library, as this breaks canonical modeling, which is a required pattern for providerhub.

File: packages/versioning/src/versioning.ts [hunk 1]
```diff
@@ -683,15 +781,23 @@ export function renamedAfter(p: Program, type: Type, version: ObjectType) {
...
   return getNameAtVersion(p, type, version) !== "";
 }
 
-export function madeOptionalAfter(p: Program, type: Type, version: ObjectType) {
+export function madeOptionalAfter(p: Program, type: Type, version: ObjectType): boolean {
   const appliesAt = appliesAtVersion(getMadeOptionalOn, p, type, version);
   return appliesAt === null ? false : !appliesAt;
 }
 
+export function hasOldTypeAtVersion(p: Program, type: Type, version: ObjectType): boolean {
```
Comment (timotheeguerin) [inline packages/versioning/src/versioning.ts:793]:
```suggestion
export function hasDifferentTypeAtVersion(p: Program, type: Type, version: ObjectType): boolean {
```

File: packages/versioning/src/versioning.ts [hunk 2]
```diff
@@ -171,6 +237,38 @@ export function getNameAtVersion(p: Program, t: Type, v: ObjectType): string {
   return "";
 }
 
+/**
+ * @returns get old type if applicable.
+ */
+export function getTypeBeforeVersion(p: Program, t: Type, v: ObjectType): any {
```
Comment (timotheeguerin) [inline packages/versioning/src/versioning.ts:243]:
```suggestion
export function getTypeBeforeVersion(p: Program, t: Type, v: ObjectType): Type | undefined {
```

File: packages/versioning/src/versioning.ts [hunk 3]
```diff
@@ -171,6 +237,38 @@ export function getNameAtVersion(p: Program, t: Type, v: ObjectType): string {
...
+  const target = toVersion(p, t, v);
+  const map = getTypeChangedFrom(p, t);
+  if (!map || !target) return "";
+
+  for (const [key, val] of map) {
+    if (target.index < key.index) {
+      return val;
+    }
+  }
+  return "";
```
Comment (timotheeguerin) [inline packages/versioning/src/versioning.ts:253]:
```suggestion
  return undefined;
```

I think those 2 make more sense here

**Review by timotheeguerin — APPROVED** [review]
A nitpick on the name and the type for the getter

File: packages/versioning/src/versioning.ts (hunk 1 above)
Comment (tjprescott) [reply packages/versioning/src/versioning.ts:793]: Fixed

File: packages/versioning/src/versioning.ts (hunk 2 above)
Comment (tjprescott) [reply packages/versioning/src/versioning.ts:243]: Fixed

File: packages/versioning/src/versioning.ts (hunk 3 above)
Comment (tjprescott) [reply packages/versioning/src/versioning.ts:253]: Fixed

**Comment by johanste** [comment]
I think it would behoove us to see what using this decorator in the context of cadl-azure-core would look like. @daviwil may have some suggestions.

**Comment by tjprescott** [comment]
@johanste in issue https://github.com/Azure/cadl-azure/issues/2405 I suggest that we make it a linter error/warning to use these in Azure specs.

**Pull request merged**